include/pybind11/detail/internals.h

```diff
@@ -314,16 +314,45 @@ inline internals **&get_internals_pp() {
     return internals_pp;
 }
 
-// forward decl
-inline void translate_exception(std::exception_ptr);
+const std::forward_list<ExceptionTranslator> &get_exception_translators();
+const std::forward_list<ExceptionTranslator> &get_local_exception_translators();
+
+// Apply all the extensions translators from a list
+// Return true if one of the translators completed without raising an exception
+// itself. Return of false indicates that if there are other translators
+// available, they should be tried.
+inline bool apply_exception_translators(const std::forward_list<ExceptionTranslator> &translators,
+                                        std::exception_ptr last_exception) {
+    for (const auto &translator : translators) {
+        try {
+            translator(last_exception);
+            return true;
+        } catch (...) {
+            last_exception = std::current_exception();
+        }
+    }
+    return false;
+}
+
+inline bool
+apply_exception_translators(const std::forward_list<ExceptionTranslator> &translators) {
+    return apply_exception_translators(translators, std::current_exception());
+}
 
 template <class T,
           enable_if_t<std::is_same<std::nested_exception, remove_cvref_t<T>>::value, int> = 0>
 bool handle_nested_exception(const T &exc, const std::exception_ptr &p) {
     std::exception_ptr nested = exc.nested_ptr();
     if (nested != nullptr && nested != p) {
-        translate_exception(nested);
-        return true;
+        const auto &local_translators = get_local_exception_translators();
```

Comment:
Do we need the temporary variables (`local_translators`, `translators`). Simply inlining the function calls would seem nicer (less code and very clear, too).